Added PE utils for sections and directories

daladim opened the pull request:
Hello,

Here are a few utility functions that I'm currently using in my own projects, but that I think everyone could benefit from, so this could make sense to upstream them :-)

philipc commented:
Thanks for the PR! I've reimplemented `ImageDataDirectory::file_offset_range` in #421 for better consistency with the existing APIs. Did you still need `Section::contains_rva` and `SectionTable::section_at` or were they only to support `ImageDataDirectory::file_offset_range`?

daladim commented:
Thanks for this other PR, that's indeed probably more suitable as what I proposed :-)

Regarding `Section::contains_rva` and `SectionTable::section_at`, I happen to use `SectionTable::section_at` in my projects (but not `contains_rva`, at least not directly).
I'd not be surprised that other people as well could need both of them one day (in case they want to do some kind of manual parsing or displaying of PE files themselves). Thus, I think that would make sense to still provide them in the public API, but that's up to you to decide (and if you want me to, I can add a commit on top your MR to implement them).

philipc commented:
They're still fine to add. You can probably update this PR, since I won't merge the other yet. Maybe rename `section_at` to something `pe_section_containing` or something like that, and rename `contains_rva` to `pe_contains_rva` (since these types are used for both PE and COFF, but these functions only apply to PE).

philipc commented on src/read/pe/section.rs:
```diff
@@ -340,6 +346,19 @@ impl pe::ImageSectionHeader {
+        match va.checked_sub(section_va) {
+            None => false,
+            Some(offset) => {
+                let (_section_offset, section_size) = self.pe_file_range();
```
Are you sure you want the file size rather than the virtual size? I think the virtual size is more likely to be what the user expects.

daladim replied:
Indeed, thanks.

daladim commented:
Sorry for the delay, I was quite busy last week.
Since I saw you've merged #421, I have rebased on 0.28.3, and I have added `Section::contains_rva` and `SectionTable::section_at`, which is quite a trivial change now.

philipc commented on src/read/pe/section.rs:
```diff
@@ -324,6 +325,11 @@ impl<'data> SectionTable<'data> {
             .find_map(|section| section.pe_data_containing(data, va))
     }
+
+    /// Return the section that contains a given virtual address in a PE file.
+    pub fn section_at(&self, va: u32) -> Option<&'data ImageSectionHeader> {
```
```suggestion
    /// Return the section that contains a given virtual address.
    pub fn section_containing(&self, va: u32) -> Option<&'data ImageSectionHeader> {
```
Still need to rename this please (sorry, I should have left it as a review comment originally). And no longer need to prefix with `pe_` like I originally requested because this is no longer using the PE file range.

daladim replied:
OK, I'm applying your suggestion.

philipc reviewed (approved):
Thanks!
